### ISSUE TITLE
controller-runtime controllers and the rest of Contour code use two different caches

### ISSUE THREAD

**Issue by skriss** [projectcontour/contour#3956]
The controller-runtime controllers (for Gateway API) use one `cache.Cache`, while the rest of Contour code uses a `cache.Cache` from `k8s.Clients` (see https://github.com/projectcontour/contour/blob/main/internal/k8s/clients.go#L36). This is both memory-inefficient, and can cause subtle race conditions due to the two caches not being 100% in sync. I've seen this manifest when a controller-runtime controller is processing an event for a new object (meaning it's in the controllers' cache), it goes to update the status of that object, and the StatusUpdater can't find that object in its cache (which is the `k8s.Clients` one) so it can't update it.

Ideally we could use one cache instance in both places. It's not immediately obvious how to wire things up correctly for this, so will require some investigation.

**Comment by skriss** [comment]
Here's a log snippet from a WIP branch for #3706 demonstrating the race condition upon creating a new `GatewayClass`:

```
INFO[0951] enqueueing gatewayclass                       context=gatewayclass-controller name=example-2
INFO[0951] reconciling gatewayclass                      context=gatewayclass-controller name=example-2
ERRO[0951] unable to update status                       context=StatusUpdateHandler error="GatewayClass.networking.x-k8s.io \"example-2\" not found" name=example-2 namespace=
```

**Comment by youngnick** [comment]
Oof, guess we all missed this one. Agreed that this should all use the same cache, definiely.

**Comment by stevesloka** [comment]
Good catch! The original idea was to move all of them over to controller runtime, but we ended up with just the Gateway bits. That might be an easier path forward to just move to that. Just a thought while thinking out loud. Would mean moving the other bits to use the same cache then.

**Comment by skriss** [comment]
Was able to get everything using the one instance of the cache and have not seen the race condition since; need to clean it up though.
